Crash if regular expression search fails for writes to addr file

To reproduce, open a new window and search for anything not in the body:
```
winid=$(9p read acme/new/ctl | awk '{print $1}')
echo /hello/ | 9p write acme/$winid/addr
```
Stack trace:
```
panic: runtime error: index out of range [0] with length 0

goroutine 61 [running]:
panic(0x634b40, 0xc000352040)
        /home/big/go/src/runtime/panic.go:706 +0x2c2 fp=0xc00015dba0 sp=0xc00015db10 pc=0x42eb22
runtime.goPanicIndex(0x0, 0x0)
        /home/big/go/src/runtime/panic.go:59 +0xa3 fp=0xc00015dbe8 sp=0xc00015dba0 pc=0x42d203
main.acmeregexp(0x0, 0x69eac0, 0xc000300118, 0xffffffffffffffff, 0xffffffffffffffff, 0x0, 0x0, 0xc000426f90, 0xe, 0x0, ...)
        /home/fhs/go/pkg/mod/github.com/rjkroege/edwood@v0.2.1-0.20190702105333-d53714b49706/addr.go:192 +0x2cd fp=0xc00015dc58 sp=0xc00015dbe8 pc=0x596e2d
main.address(0xc000300000, 0x69eac0, 0xc000300118, 0xffffffffffffffff, 0xffffffffffffffff, 0x0, 0x0, 0x0, 0x12, 0xc00015df18, ...)
        /home/fhs/go/pkg/mod/github.com/rjkroege/edwood@v0.2.1-0.20190702105333-d53714b49706/addr.go:316 +0x7a2 fp=0xc00015dd80 sp=0xc00015dc58 pc=0x5975e2
main.xfidwrite(0xc0001fe6e0)
        /home/fhs/go/pkg/mod/github.com/rjkroege/edwood@v0.2.1-0.20190702105333-d53714b49706/xfid.go:510 +0x606 fp=0xc00015df88 sp=0xc00015dd80 pc=0x5d1dd6
main.xfidctl(0xc0001fe6e0, 0x69f3a0, 0xc00009a130)
        /home/fhs/go/pkg/mod/github.com/rjkroege/edwood@v0.2.1-0.20190702105333-d53714b49706/xfid.go:52 +0x9c fp=0xc00015dfc8 sp=0xc00015df88 pc=0x5cf78c
runtime.goexit()
        /home/big/go/src/runtime/asm_amd64.s:1357 +0x1 fp=0xc00015dfd0 sp=0xc00015dfc8 pc=0x45c101
created by main.xfidallocthread
        /home/fhs/go/pkg/mod/github.com/rjkroege/edwood@v0.2.1-0.20190702105333-d53714b49706/acme.go:619 +0x1d4
```

--- a/addr.go
+++ b/addr.go
@@ -2,10 +2,12 @@ package main
 
 import "strings"
 
+// These constants indicates the direction of regular expresssion search.
+// TODO(fhs): Introduce a new type for these constants.
 const (
 	None = iota
-	Fore = '+'
-	Back = '-'
+	Fore = '+' // Forward
+	Back = '-' // Backward
 )
 
 const (
@@ -152,12 +154,15 @@ Rescue:
 
 var pattern *AcmeRegexp
 
-func acmeregexp(showerr bool, t Texter, lim Range, r Range, pat string, dir int) (retr Range, foundp bool) {
-	var (
-		sel RangeSet
-		q   int
-		err error
-	)
+// acmeregexp searches for regular expression pattern pat in text t.
+// If pat is empty, it uses the pattern used in the previous invocation of this function.
+// Dir indicates the direction of the search: forward or backward.
+// R sets the text position where search begins.
+// Lim sets the text position where search ends for forward search
+// (set range to {-1, -1} for no limit).
+// Warnings will be shown to user if showerr is true.
+// It returns the match and whether a match was found.
+func acmeregexp(showerr bool, t Texter, lim Range, r Range, pat string, dir int) (retr Range, found bool) {
 	if len(pat) == 0 && pattern == nil {
 		if showerr {
 			warning(nil, "no previous regular expression\n")
@@ -165,17 +170,19 @@ func acmeregexp(showerr bool, t Texter, lim Range, r Range, pat string, dir int)
 		return r, false
 	}
 	if len(pat) > 0 {
+		var err error
 		pattern, err = rxcompile(pat)
 		if err != nil {
 			return r, false
 		}
 	}
+
+	var sel RangeSet
 	if dir == Back {
 		sel = pattern.rxbexecute(t, r.q0, 1)
 	} else {
-		if lim.q0 < 0 {
-			q = -1
-		} else {
+		q := -1
+		if lim.q0 >= 0 {
 			q = lim.q1
 		}
 		sels := pattern.rxexecute(t, nil, r.q1, q, 1)
@@ -185,29 +192,35 @@ func acmeregexp(showerr bool, t Texter, lim Range, r Range, pat string, dir int)
 			sel = nil
 		}
 	}
-	if len(sel) == 0 && showerr {
-		warning(nil, "no match for regexp\n")
+	if len(sel) == 0 {
+		if showerr {
+			warning(nil, "no match for regexp\n")
+		}
 		return Range{-1, -1}, false
 	}
 	return sel[0], true
 }
 
-// getc takes a closure over the address expression and returns the qth rune.
+// address parses an address for text t, where getc takes a closure over
+// the address expression and returns the qth rune in the address
+// (q0 <= q < q1). If eval is true, the address is also evaluated.
+// Lim sets the limits of regular expression search.
+// Warnings will be shown to user if showerr is true.
+// It returns the updated address range (initially set to ar), whether
+// the evaluation was successful, and the position q in the address
+// where parsing was stopped.
 func address(showerr bool, t Texter, lim Range, ar Range, q0 int, q1 int, getc func(q int) rune, eval bool) (r Range, evalp bool, qp int) {
 	var (
-		dir, size    int
-		n            int
-		prevc, c, nc rune
-		q            int
-		pat          string
-		nr           Range
+		n         int
+		prevc, nc rune
+		nr        Range
 	)
 	evalp = eval
 	r = ar
-	q = q0
-	dir = None
-	size = Line
-	c = 0
+	q := q0
+	dir := None
+	size := Line
+	c := rune(0)
 	for q < q1 {
 		prevc = c
 		c = getc(q)
@@ -291,7 +304,7 @@ func address(showerr bool, t Texter, lim Range, ar Range, q0 int, q1 int, getc f
 			dir = Back
 			fallthrough
 		case c == '/':
-			pat = ""
+			pat := ""
 			for q < q1 {
 				c = getc(q)
 				q++

--- a/addr_test.go
+++ b/addr_test.go
@@ -2,6 +2,7 @@ package main
 
 import (
 	"fmt"
+	"reflect"
 	"testing"
 )
 
@@ -57,6 +58,68 @@ func TestAddr(t *testing.T) {
 			if test.r != r || test.ep != ep || test.q != q {
 				t.Errorf("address %q: r=%v, ep=%v, q=%v; expected r=%v, ep=%v, q=%v",
 					test.addr, r, ep, q, test.r, test.ep, test.q)
+			}
+		})
+	}
+}
+
+func TestAcmeregexp(t *testing.T) {
+	warnings = nil
+	pattern = nil
+	defer func() {
+		warnings = nil
+		pattern = nil
+	}()
+
+	tt := []struct {
+		name    string
+		pat     string
+		dir     int
+		r       Range
+		found   bool
+		warning string
+	}{
+		{"FirstEmptyPat", "", Fore, Range{0, 0}, false, "no previous regular expression\n"},
+		{"pat=abcd", "abcd", Fore, Range{0, 4}, true, ""},
+		{"SecondEmptyPat", "", Fore, Range{0, 4}, true, ""},
+		{"pat=αβξδ", "αβξδ", Fore, Range{5, 9}, true, ""},
+		{"NoMatch", "xyz", Fore, Range{-1, -1}, false, "no match for regexp\n"},
+		{"InvalidPat", "(abcd", Fore, Range{0, 0}, false, ""},
+		{"Backwards", "abcd", Back, Range{0, 4}, true, ""},
+	}
+
+	for _, tc := range tt {
+		t.Run(tc.name, func(t *testing.T) {
+			warnings = nil
+			text := &Text{
+				file: &File{
+					b: Buffer([]rune("abcd αβξδ\n")),
+				},
+			}
+			lim := Range{
+				0,
+				text.file.Nr(),
+			}
+			start := Range{0, 0}
+			if tc.dir == Back {
+				start = Range{text.file.Nr(), text.file.Nr()}
+			}
+			r, found := acmeregexp(true, text, lim, start, tc.pat, tc.dir)
+			if found != tc.found {
+				t.Errorf("found=%v; want %v", found, tc.found)
+			}
+			if !reflect.DeepEqual(r, tc.r) {
+				t.Errorf("range=%v; want %v", r, tc.r)
+			}
+			if tc.warning != "" {
+				want := tc.warning
+				if len(warnings) == 0 {
+					t.Fatalf("no warning generated; want %q", want)
+				}
+				got := string([]rune(warnings[0].buf))
+				if got != want {
+					t.Errorf("warning is %q; want %q", got, want)
+				}
 			}
 		})
 	}

--- a/xfid.go
+++ b/xfid.go
@@ -426,18 +426,16 @@ func fullrunewrite(x *Xfid) []rune {
 	return r
 }
 
+// xfidwrite responds to a plan9.Twrite request.
 func xfidwrite(x *Xfid) {
 	// log.Println("xfidwrite", x)
 	// defer log.Println("done xfidwrite")
 	var (
-		fc               plan9.Fcall
-		c                int
-		eval             bool
-		r                []rune
-		a                Range
-		t                *Text
-		q0, tq0, tq1, nb int
-		err              error
+		fc           plan9.Fcall
+		c            int
+		t            *Text
+		q0, tq0, tq1 int
+		err          error
 	)
 	qid := FILE(x.f.qid)
 	w := x.f.w
@@ -503,13 +501,13 @@ func xfidwrite(x *Xfid) {
 		x.respond(&fc, nil)
 
 	case QWaddr:
-		r = []rune(string(x.fcall.Data))
+		r := []rune(string(x.fcall.Data))
 		t = &w.body
 		w.Commit(t)
-		eval = true
-		a, eval, nb = address(false, t, w.limit, w.addr, 0, len(r),
+		eval := true
+		a, eval, nr := address(false, t, w.limit, w.addr, 0, len(r),
 			func(q int) rune { return r[q] }, eval)
-		if nb < len(r) {
+		if nr < len(r) {
 			x.respond(&fc, ErrBadAddr)
 			break
 		}
@@ -524,7 +522,7 @@ func xfidwrite(x *Xfid) {
 	case Qeditout:
 		fallthrough
 	case QWeditout:
-		r = fullrunewrite(x)
+		r := fullrunewrite(x)
 		if w != nil {
 			err = edittext(w, w.wrselrange.q1, r)
 		} else {
@@ -552,7 +550,7 @@ func xfidwrite(x *Xfid) {
 		xfidctlwrite(x, w)
 
 	case QWdata:
-		a = w.addr
+		a := w.addr
 		t = &w.body
 		w.Commit(t)
 		if a.q0 > t.Nc() || a.q1 > t.Nc() {

--- a/xfid_test.go
+++ b/xfid_test.go
@@ -77,6 +77,53 @@ func (mr *mockResponder) respond(x *Xfid, t *plan9.Fcall, err error) *Xfid {
 
 func (mr *mockResponder) msize() int { return 8192 }
 
+func TestXfidwriteQWaddr(t *testing.T) {
+	for _, tc := range []struct {
+		name string
+		addr []byte
+		r    Range
+		err  error
+	}{
+		{"ErrAddrRange", []byte("/hello/"), Range{}, ErrAddrRange},
+		{"ErrBadAddr", []byte("/hello/\n"), Range{}, ErrBadAddr},
+		{"ValidAddr", []byte("/cα/"), Range{2, 4}, nil},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			mr := new(mockResponder)
+			w := NewWindow().initHeadless(nil)
+			w.body.file.b = Buffer([]rune("abcαβξ\n"))
+			w.col = new(Column)
+			w.limit = Range{0, w.body.file.Nr()}
+			x := &Xfid{
+				fcall: plan9.Fcall{
+					Data: []byte(tc.addr),
+				},
+				f: &Fid{
+					qid: plan9.Qid{
+						Path: QID(0, QWaddr),
+					},
+					w: w,
+				},
+				fs: mr,
+			}
+			xfidwrite(x)
+			if mr.err != tc.err {
+				t.Fatalf("error is %v; want %v", mr.err, tc.err)
+			}
+			if mr.err == nil {
+				got := mr.fcall.Count
+				want := uint32(len(tc.addr))
+				if got != want {
+					t.Errorf("fcall.Count is %v; want %v", got, want)
+				}
+				if !reflect.DeepEqual(w.addr, tc.r) {
+					t.Errorf("window address is %v; want %v", w.addr, tc.r)
+				}
+			}
+		})
+	}
+}
+
 func TestXfidruneread(t *testing.T) {
 	tt := []struct {
 		body   []rune // window body

